### PR TITLE
drawCuboid, drawLineCuboid(), and related functions

### PR DESCRIPTION
**Pull request by NinjolasNJM**
- `cuboid` type is now included and used in both TextureMap and Generator_Builder.
- `makeCuboid()` - cuboid constructor that takes a source position and height, length and width.
- TextureMap is simplified by using `makeCuboid()`
- `drawCuboid()` - draws each side of a cuboid, and can draw either a left or right sided cuboid.

- `fillRect()` - draws a solid rectangle of a specified color.

- `drawLinePath()` - draws a path of lines given an array of points.
- `drawLineRect()` - draws a rectangular path given a rectangle.
- `drawLineCuboid()` - draws lines that surround a cuboid destination, and can be left or right sided.
- `drawFoldLinePath(), drawFoldLineRect(), and drawFoldLineCuboid()` - draw paths, rectangles, and cuboids respectively with fold lines.

The Character Generator and Mini Character Generator now use the new cuboid and line related functions, and the Mini Character Generator also now has a new title and opaque background.

### CODE REVIEW

File: src/builder/modules/Generator.res [hunk 1]
```diff
@@ -130,6 +130,10 @@ let fillBackgroundColorWithWhite = () => {
   model := Generator_Builder.fillBackgroundColor(model.contents, "#ffffff")
 }
 
+let fillRect = (dest: Builder.rectangle, color: string) => {
```
Comment (pixelpapercraft) [inline src/builder/modules/Generator.res:133]: Hi @NinjolasNJM for each of these functions, let's justify their purpose in the context of papercraft designs.

For `fillRect()` I can see this might be useful as a background fill behind things that are white (so we can see the edges), perhaps? Any other uses you have in mind?

File: src/builder/modules/Generator.res [hunk 2]
```diff
@@ -143,6 +147,72 @@ let drawLine = (
     Generator_Builder.drawLine(model.contents, from, to, ~color, ~width, ~pattern=[], ~offset=0)
 }
 
+let drawLinePath = (
```
Comment (pixelpapercraft) [inline src/builder/modules/Generator.res:150]: Since we already have `drawLine()` and `drawLinePath()` is just an alternative to that, I'm not seeing a compelling need for this function? Keeping the generator as simple as possible is a fundamental principle to help us decide what gets included, and less functions is a key factor for simplicity. Maybe we use it in the Generator_Builder for some purpose, but I'm not sure it needs to be a feature exposed to generator writers?

File: src/builder/modules/Generator.res [hunk 3]
```diff
@@ -143,6 +147,72 @@ let drawLine = (
...
+      points,
+      ~color,
+      ~width,
+      ~pattern=[],
+      ~offset=0,
+      ~close,
+    )
+}
+
+let drawLineRect = (
```
Comment (pixelpapercraft) [inline src/builder/modules/Generator.res:171]: I'm guessing this one could be used to "section off" parts of a design? Maybe `drawLine()` would be good enough for that? And particular use cases you can think of for this one?

File: src/builder/modules/Generator.res [hunk 4]
```diff
@@ -143,6 +147,72 @@ let drawLine = (
...
+      [(x, y), (x + w, y), (x + w, y + h), (x - 1, y + h), (x - 1, y)],
+      ~color,
+      ~width,
+      ~pattern=[],
+      ~offset=0,
+      ~close=false,
+    )
+}
+
+let drawLineCuboid = (
```
Comment (pixelpapercraft) [inline src/builder/modules/Generator.res:192]: Drawing cuboids is a very interesting idea! Is there any particular use case you have in mind for drawing a line cuboid?

File: src/builder/modules/Generator.res (hunk 1 above)
Comment (NinjolasNJM) [reply src/builder/modules/Generator.res:133]: the fillRect() function I made so that the mini generator could have the faces of the head that need to be glued on highlighted the same way they were in the original design- similar uses could be to eliminate the need for images on other colored labels.

File: src/builder/modules/Generator.res (hunk 2 above)
Comment (NinjolasNJM) [reply src/builder/modules/Generator.res:150]: drawLinePath() isn't so much an alternative to drawLine() as it is a shorthand for strings of connected lines- think for outlines of complex shapes. It is also used in the drawRect() function, which in turn is used in drawCuboid().

File: src/builder/modules/Generator.res [hunk 5]
```diff
@@ -156,6 +226,53 @@ let drawFoldLine = (from: Builder.position, to: Builder.position) => {
     )
 }
 
+let drawFoldLinePath = (points: array<Builder.position>, ~close: bool=false, ()) => {
```
Comment (pixelpapercraft) [inline src/builder/modules/Generator.res:229]: Similar to my comment about the `drawLine()` I wonder if there is really much need for two similar fold line functions? This is marginally nicer to use for multiple joined lines, but it's still trivial to do the equivalent with `drawFoldLine()`.

File: src/builder/modules/Generator.res (hunk 4 above)
Comment (NinjolasNJM) [reply src/builder/modules/Generator.res:192]: This and drawLineRect() are mostly for consistency, so that the functions don't only exist for fold lines, but a specific use could be possibly for color coding similar looking parts with outlines, such as for the squid. these could go underneath the fold lines and still be visible through the dashes, and both could be toggleable on or off.

File: src/builder/modules/Generator.res (hunk 5 above)
Comment (NinjolasNJM) [reply src/builder/modules/Generator.res:229]: The path functions will be very useful for drawing fold outlines, since the marching squares algorithm for finding the outline will return an array of positions, and the drawPath() function will be able to take those as a parameter.

File: src/builder/modules/Generator.res [hunk 6]
```diff
@@ -156,6 +226,53 @@ let drawFoldLine = (from: Builder.position, to: Builder.position) => {
...
+      points,
+      ~color="#7b7b7b",
+      ~width=1,
+      ~pattern=[2, 2],
+      ~offset=3,
+      ~close,
+    )
+}
+
+let drawFoldLineRect = (dest: Builder.rectangle) => {
```
Comment (pixelpapercraft) [inline src/builder/modules/Generator.res:242]: I believe your goal is to create a great experience by providing many choices to choose from - and thanks for these creative suggestions.

The consequence is the builder becomes significantly less simple when there are multiple versions of essentially the same thing. Unless there is a genuinely compelling reason to keep a function, let's omit it.

File: src/builder/modules/Generator.res [hunk 7]
```diff
@@ -156,6 +226,53 @@ let drawFoldLine = (from: Builder.position, to: Builder.position) => {
...
+      [(x, y), (x + w, y), (x + w, y + h), (x - 1, y + h), (x - 1, y)],
+      ~color="#7b7b7b",
+      ~width=1,
+      ~pattern=[2, 2],
+      ~offset=3,
+      ~close=false,
+    )
+}
+
+let drawFoldLineCuboid = (
```
Comment (pixelpapercraft) [inline src/builder/modules/Generator.res:256]: This one is a very interesting idea.

Your recent suggestions have led me to also think about having a `drawTab()` function.

This also leads me to think about where the fold lines belong.

For example, it could be something like this:

[1] `drawFoldLineCuboid()` only draws the internal fold lines, and no edge/external fold lines.
[2] `drawTab()` draws a fold line at its base, which then becomes an edge fold on a cuboid.

These days I am wondering about the value of having fold lines along the outsides of designs, since no fold is actually needed there.

Let me know your thoughts on the above.

File: src/builder/modules/Generator.res [hunk 8]
```diff
@@ -199,6 +316,34 @@ let drawTextureLegacy = (
   )
 }
 
+let drawCuboid = (
```
Comment (pixelpapercraft) [inline src/builder/modules/Generator.res:319]: This is a great helper function.

I'm thinking that this one actually belongs in a new module rather than in the builder, since it's specific to Minecraft textures (specifically how it orients it's texture patterns).

I am thinking we create a new module next to the `TextureMap.res` module and name it `Minecraft.res` and let's move this function over there.

What do you think?


I am thinking we will move some of the code in `TextureMap.res` into this new module as well, but that change is for a separate PR.

File: src/builder/modules/Generator.res (hunk 7 above)
Comment (NinjolasNJM) [reply src/builder/modules/Generator.res:256]: Some of the exterior "fold" lines actually are still folded along even if they are not the border of two faces- they are still folded when the tabs are folded. The rest of the exterior folds on the cuboid still represent *where* a fold should be, and though they aren't where a fold is actually made, they can help with aligning where the edges of the paper are.

drawTab() is something that would be great thing to have, but I want to see if the antialiasing issue can be dealt with first before adding it. As for having the fold for that edge be part of the drawTav() function, this would probably add unnecessary complexity, as most generators have toggleable folds, meaning the function would also need a boolean for whether to draw one or not. In a case like this, ~drawFolds and ~drawTabs parameters might as well be added to the drawCuboid() function instead, so that the texture, tabs and folds could be drawn in one step.

File: src/builder/modules/Generator_Builder.res [hunk 9]
```diff
@@ -41,6 +41,24 @@ type rectangleLegacy = {
...
+type cuboid = {
+  top: rectangle,
+  bottom: rectangle,
+  front: rectangle,
+  right: rectangle,
+  left: rectangle,
+  back: rectangle,
+}
+
+let makeCuboid = ((x, y): position, (w, h, l)): cuboid => {
```
Comment (pixelpapercraft) [inline src/builder/modules/Generator_Builder.res:53]: Nice! Small suggestion - both width and height are also "lengths", so could I suggest we use width, height, and _depth_.

Also, I did a quick diagram - can you check if this is correct and add it as a comment above this function?

```
W = Width
H = Height
D = Depth

(x,y)            
  .         +----W----+    
            |         |
            D   Top   D 
            |         |
  +----D----+----W----+----D----+----W----+
  |         |         |         |         |
  H  Right  H  Front  H  Left   H  Back   H
  |         |         |         |         |
  +----D----+----W----+----D----+----W----+
            |         |
            D Bottom  D
            |         |
            +----W----+    
```

A few thoughts:

I'm thinking about separate "creating" cuboids from "positioning" cuboids.

For example:
```
// Maybe move into a module
module Cuboid = {
 type t = ... // cuboid type goes here
  let make = (width: int, height: int, depth: int): t => { } // Creates a new cubiod, independent of position.
  let translate = (cuboid: t, x: int, y: int): t => { } // Translates the cuboid by the given x and y coordinates.
} 
```

Example usage:

```res
Cuboid.make(3, 12, 4)->Cuboid.translate(40, 16)
```

Not sure, just thinking about it at the moment.

File: src/builder/modules/Generator_Builder.res [hunk 10]
```diff
@@ -457,8 +490,13 @@ let getOffset = ((x1, y1): position, (x2, y2): position) => {
...
   in a fully opaque line with the correct width if the line is vertical or
   horizontal, but antialiasing may still affect lines at other angles.
+
+  The angle should be 3 pi / 2 radians if the angle is 270 degrees, but with 
+  the ternary making it pi / 2 regardless of sign the rectangle lines functions
+  are much easier to do, for some reason. Maybe later, this could be replaced 
+  with something using the absolute value of the angle?
  */
-  let angle = Js.Math.atan2(~y=h, ~x=w, ())
+  let angle = w === 0.0 ? Js.Math._PI /. 2.0 : Js.Math.atan2(~y=h, ~x=w, ())
```
Comment (pixelpapercraft) [inline src/builder/modules/Generator_Builder.res:499]: It would be good to think about this one some more.

In the current code, you can control if a line is offset to the "left" or offset to the "right" by just changing the direction the line is drawn, e.g. instead of A to B, draw as B to A.

With this change using `w === 0`, it means that vertical lines are always offset to one side only independent of the point order, however all other lines continue to have the previous offset behaviour (adjusting depending on the order of the points), so the `w ===0` is inconsistent with all lines where `w !== 0`.

For now I'm thinking we leaving this as-is and change any calling code to adjust it's point drawing order to get the desired result.

File: src/builder/modules/Generator.res (hunk 1 above)
Comment (pixelpapercraft) [reply src/builder/modules/Generator.res:133]: > so that the mini generator could have the faces of the head that need to be glued on highlighted the same way they were in the original design

Are you referring to this part?

image

Could you explain the idea a bit more (I forgot what did on the original design)?

> images on other colored labels.

That's a great example. Actually we could probably benefit from a `drawLabel()` function with an optional background colour eventiually.

File: src/builder/modules/Generator.res (hunk 2 above)
Comment (pixelpapercraft) [reply src/builder/modules/Generator.res:150]: Cool, thanks for explaining.

I do agree that this is a slightly nicer way to handle drawing a line with many points. So the papercraft use case here is to draw a line for a shape with many points (such as a complex shape as you mentioned).

Just a side question - I can't recall a use case for drawing solid line outlines? But you might have something in mind?

Back to the main idea - since we would need an array of points anyway for the shape, we could also just loop over those points and call `drawLine()` right? A loop is a little less convenient, but still trivial to write. And if this use case is rare, then a loop with `drawLine()` should be sufficent?
 
> It is also used in the drawRect() function, which in turn is used in drawCuboid().

`drawRect()` could easily be switched across to use `drawLine()` right? And `drawCuboid()` would not need to change? Is that right? That "offset" chat in the drawLine function might also be relevant here.

So here's my understanding, but correct me if this is not quite right:

* Primary use case is shapes with many points (a few points, such as 4, are easy enough t.o do with drawLine(), so perhaps this helps with shapes with > 4 points), however this use case is not very common.
* This is a little more convenient than `drawLine()` since looping is not required, but `drawLine()` in a loop would achieve the same result, and is trivial to do.

And my other comment:

* I can't recall any use cases for drawing solid outlines, but let me know if you have some in mind.

File: src/builder/modules/Generator.res (hunk 4 above)
Comment (pixelpapercraft) [reply src/builder/modules/Generator.res:192]: > This and drawLineRect() are mostly for consistency, so that the functions don't only exist for fold lines

Yes, I had a feeling that might have been the case.

> but a specific use could be possibly for color coding similar looking parts with outlines, such as for the squid. these could go underneath the fold lines and still be visible through the dashes, and both could be toggleable on or off.

I hadn't thought of that, great creative thinking. I understand that's just an example, but for that case it's probably an improvement we could make to the `drawFoldLine()` function - standardising how coloured fold lines might behave.

File: src/builder/modules/Generator.res (hunk 5 above)
Comment (pixelpapercraft) [reply src/builder/modules/Generator.res:229]: For an outlining feature, that path drawing will need to occur with the Generator_Builder (it won't be possible for the Generator_Builder to call functions on the Generator module).

The pros and cons here are probably similar to the line drawing function chat above. I.e. potentially has benefit when there are many points, but could also be achieved easily with `drawFoldLine()` within a loop?

File: src/builder/modules/Generator.res (hunk 7 above)
Comment (pixelpapercraft) [reply src/builder/modules/Generator.res:256]: > Some of the exterior "fold" lines actually are still folded along even if they are not the border of two faces- they are still folded when the tabs are folded.

👍 

> The rest of the exterior folds on the cuboid still represent where a fold should be, and though they aren't where a fold is actually made, they can help with aligning where the edges of the paper are.

Can you explain a bit more about the aligning idea?

> drawTab() is something that would be great thing to have, but I want to see if the antialiasing issue can be dealt with first before adding it.

TBH I am not concerned with the antialiasing here. If I had to choose between generated antialiased lines in automatic vs pixel perfect tabs in images used as overlays, I would definitely favour the generated version - it's a great improvement over the image strategy.

We could implement our own line drawing function which is not antialiased, as you can image, it can also look a bit rough depending on the angle.

> As for having the fold for that edge be part of the drawTav() function.

You've raised some good points here.

There are some implementation details to work out, but here's the concept I have in mind - we only draw a fold line where you literally need to do a fold. We don't draw fold lines on edges where there are no tabs because there is no fold (i.e. a join is not a fold).

This is pretty easy to implement, and should produce a nice consistent pattern we can apply to all generators. What do you think?

File: src/builder/modules/Generator.res (hunk 1 above)
Comment (NinjolasNJM) [reply src/builder/modules/Generator.res:133]: Yes- in the original design, those faces were like this:

In this PR, I added this change to the mini generator, along with having it use the new line and fold functions.

>  Actually we could probably benefit from a drawLabel() function with an optional background colour eventiually.

This is a good idea for a function that could work well with fleshing out the text drawing function- after this PR, its use could be expanded, and maybe a sign generator could be made using it.

File: src/builder/modules/Generator.res (hunk 5 above)
Comment (NinjolasNJM) [reply src/builder/modules/Generator.res:229]: One of the reasons for using drawPath() over drawLine() in a loop is that drawPath() renders once when it is called- it isn't known how much this could affect rendering times, but it should be a good idea to draw as few times as needed, especially if generators start drawing folds, tabs, and labels each during multiple steps.

File: src/builder/modules/Generator.res (hunk 7 above)
Comment (NinjolasNJM) [reply src/builder/modules/Generator.res:256]: Having outlines even where there isn't a fold has a few benefits over not having them:

- It aids with cutting and aligning edges- When cutting along the very edge of a face, a dashed line can help with making a straighter cut. It gives both an easily visible and more forgiving border that can be seen better than the border between the blank and textured parts of the paper, especially when the paper is farther from view if it is cut with a knife on a desk, or when the texture is white or close to white. When gluing edge to a tab, this visual advantage remains, and can both help with making a cuboid as straight as possible, and with making any slight errors less obvious.
- It makes designing with different tab placements easier- Since sometimes tabs need to be put on edges other than where they are usually, if outside folds are only on tabs, then this adds complexity to where they are drawn. drawFoldLineCuboid() only draws two rectangles and a line and draws on every edge, but not drawing on bare edges increases this to 6 calls of drawFoldLine() at the simplest, and needing to account for whether a tab is present means individually drawing along every edge, defeating the function's purpose. It would be much simpler to treat every edge the same.
- Having fold lines as edge lines is also a much more common standard to be used than having just the folded edges given dashed lines, especially after solid red outlines fell out of favor because of their aesthetics. As opposed to adjusting or creating new standards for designs, I think that the generator functions should instead be to perpetuate and enable existing conventions, making them more attractive to use and discouraging designers from creating new ones that conflict with existing ones.

File: src/builder/modules/Generator.res (hunk 8 above)
Comment (NinjolasNJM) [reply src/builder/modules/Generator.res:319]: That does make sense to do, but what would be in `TextureMap.res` and what would be in `Minecraft.res`?

File: src/builder/modules/Generator_Builder.res (hunk 9 above)
Comment (NinjolasNJM) [reply src/builder/modules/Generator_Builder.res:53]: > Nice! Small suggestion - both width and height are also "lengths", so could I suggest we use width, height, and depth.

From what I've gathered, "height", "width", "depth", "breadth", and "length" can be used interchangeably, but in the cases for the dimensions of a box, such as a papercraft cuboid, I've seen width and length be used for the horizontal axes, and either height or depth be used for the vertical axis. However, it is minor, and width and length could get easily confused for each other, while "depth" might feel more distinct from the other two. I'll think about this.

> I'm thinking about separate "creating" cuboids from "positioning" cuboids.

I think this is the solution I was looking for earlier- I wanted to have makeCuboid() be just the constructor for the cuboid type, but I couldn't figure out what that would look like while still having cuboid be a record of rectangles. I also wanted to have cuboids be used for the destination in drawCuboid(), but I ended up not doing so since there was not preexisting map for body parts. These might be added to `Minecraft.res` when it is made.

**Comment by pixelpapercraft** [comment]
Hi @NinjolasNJM 

You've made some good comments about the line strategy, so I took a look at how papercraft designs are made on other sites. Solid outlines were on all of the examples I saw. Fold lines were either solid or dashed.

I've applied these ideas to this example:

image

Concept is simple:

* Solid line for where the cuts occur.
* Dashed line for where the folds occur.

Couple of points:

* The cut lines are our standard grey. Lots of online examples used black, but I didn't change to that here.
* The fold lines follow a 4-4 pattern rather than a 2-2 pattern. I don't recall seeing any examples from other sites with with the smaller fold pattern.

I really like the simplicity of the concept - we can easily apply it everywhere, and it looks good to me. What do you think?

Actually would be great to get @PaperDoggy in on this chat as well, and get any other thoughts.

**Comment by NinjolasNJM** [comment]
The topic of other sites raises how other kinds of papercraft are different to what we usually have, and how the differences in their styles and ours reflect these:
- Other papercraft is usually at a much larger scale than the 800% texture scale we use, and often are design with no printed size in mind- this is why a solid line is used over a possibly hard to see dash line, when dash patterns are used, they tend to be larger.
- Other papercraft usually uses models and textures made for the papercraft design as opposed to existing game textures, and as such faces are often a solid color or have minimal texturing, meaning a solid fold or cut line is less obtrusive as there is no texture to disturb.
- Other papercrafts are often vector-based and/or use non-cuboids with many diagonal lines, which might make dashed lines look less clean when or if the image is rasterized for printing.

Like what I said earlier about changing standards, I don't think that the functions in the generator builder should try to create new standards, and instead should define and solidify what we've already been using. In the end, conventions like these aren't supposed to correct trends that are of little consequence, but rather enable them.

**Comment by pixelpapercraft** [comment]
Thanks for that perspective @NinjolasNJM

I'll do my best to address your points.

> Other papercraft is usually at a much larger scale than the 800% texture scale we use, and often are design with no printed size in mind- this is why a solid line is used over a possibly hard to see dash line, when dash patterns are used, they tend to be larger.

Some designs on other sites use solid lines for fold lines. As you said, it could be because the designer perceives dashed lines as harder to see, but I'd be surprised if that was actually the case, because the designer could just make them thicker (I'm assuming vector designs here). IMO it's more likely to be simply a stylistic decision. Unless, you were referring to print resolution? In that case it's not really something the designer can control for.

Most of the other designs I saw use solid lines for where to cut and dashed lines for where to fold. This seems like a pretty standard and simple convention.

> Other papercraft usually uses models and textures made for the papercraft design as opposed to existing game textures, and as such faces are often a solid color or have minimal texturing, meaning a solid fold or cut line is less obtrusive as there is no texture to disturb.

If I understand correctly, your concern is that a solid line may interfere with the textures. In the proposal above, the cut lines are only on the outside. And if there is a need for a cut line within a texture, we will need to represent that in some way, and a solid line is what we do today, right?

> Other papercrafts are often vector-based and/or use non-cuboids with many diagonal lines, which might make dashed lines look less clean when or if the image is rasterized for printing.

I might be misunderstanding the idea here, but vector based designs don't need to be rasterized for printing, they can have perfect clean printed lines independent of the angle. Also dashed lines are still common in those designs with many diagonal lines, but used exclusively as fold lines.

> Like what I said earlier about changing standards, I don't think that the functions in the generator builder should try to create new standards, and instead should define and solidify what we've already been using. In the end, conventions like these aren't supposed to correct trends that are of little consequence, but rather enable them.

I think here you are referring to continuing with "Pixel Papercraft standards", is that right?

Something I have in mind is that we should be aiming to cater to the global papercraft community, not just the existing Pixel Papercraft community with conventions we've applied in the past. And TBH, we never really settled on any particular conventions, we have a few different ideas depending on which papercraft model you look at. That's why it's great to have this chat now.

It would be great if a random new person arrives at the site, and the conventions they've seen on other sites are also present here.

Lastly, I'd consider Alsa Honggo to be one of the best designers recently on the site, so Alsa's designs are good for us to consider, for example:

https://www.pixelpapercraft.com/papercraft/60e36ea04273130c62d01654/goat

So short story, let's use this as an opportunity to refine what we are currently doing, and use standards that are used elsewhere by other professional papercraft designers.

And to recap, the idea is simple:

* Cut lines are solid lines
* Fold lines are dashed lines 

This is easy for everyone to understand, and apply in generator code.

But let me know if you feel I might be missing something important. Thanks!

File: src/builder/modules/Generator.res (hunk 5 above)
Comment (pixelpapercraft) [reply src/builder/modules/Generator.res:229]: > it isn't known how much this could affect rendering times

As long as we're not drawing hundreds/thousands the there should be no noticeable difference (in the order of a few milliseconds) so it's not something to be concerned about for now.

Something else that I suspect will come up is that line _offset_ matter.

* With drawPath() we won't have control of the offset of each line segment.
* With drawLine() we will have control of the offset of each line (i.e. reversing the direction of the line drawing changes which side the offset is applied to).

There are some reasons what drawLine() will be favourable here, but let me know if you'd like me to explain further.

File: src/builder/modules/Generator.res (hunk 7 above)
Comment (pixelpapercraft) [reply src/builder/modules/Generator.res:256]: Some great points here.

Once we make a call on what we want the end result to look like, it will help guide decisions on what functions we need and how they should behave.

File: src/builder/modules/Generator.res (hunk 8 above)
Comment (pixelpapercraft) [reply src/builder/modules/Generator.res:319]: I suspect that probably everything will move across into `Minecraft.res` - I probably should have named it that from the beginning.

**Comment by NinjolasNJM** [comment]
> As you said, it could be because the designer perceives dashed lines as harder to see, but I'd be surprised if that was actually the case, because the designer could just make them thicker (I'm assuming vector designs here). IMO it's more likely to be simply a stylistic decision. Unless, you were referring to print resolution? In that case it's not really something the designer can control for.
What I mean is that when a one in-game pixel wide edge has a dash line with a 4-4 pattern on it, it only has a single dashed part drawn on that edge, and if it is smaller then it may not have anything drawn on it at all. Having a smaller pattern means that smaller edges, like the ones often seen in our kinds of papercraft, can be more likely to have their edge drawn on.

> If I understand correctly, your concern is that a solid line may interfere with the textures. In the proposal above, the cut lines are only on the outside. And if there is a need for a cut line within a texture, we will need to represent that in some way, and a solid line is what we do today, right?
The concern here is how the design will look once completed. An edge with a solid border will look different from one with just a dashed one, and this breaks the symmetry that the cuboid will have in regards to its edges- Imagine if the PPC logo had one of the edges solid while the other two are still dashed.

> Something I have in mind is that we should be aiming to cater to the global papercraft community, not just the existing Pixel Papercraft community with conventions we've applied in the past. And TBH, we never really settled on any particular conventions, we have a few different ideas depending on which papercraft model you look at. That's why it's great to have this chat now.
If we want to cater to the greater papercraft community, I don't think that becoming more like average papercraft sites would be that good of an idea- other sites seem to be more disorganized both in their design conventions and with their users ( often there are no users or forum aspect- they are just personal or corporate blogs) and appealing to that sentiment ignores how this community *is* one of the tightest and most charismatic papercraft communities today, and how expansion should come about by taking the community's uniquities outside its original scope, not the other way around.

> Some designs on other sites use solid lines for fold lines.
From what I've researched, it seems that most designs use solid lines for fold lines- and have no distinction between fold and cut lines. Out of the first 50 papercraft designs on Google Images, excluding the ones from pixel papercraft, there were 31 that made no distinction between cut and fold lines on edges- either using solid lines, or no lines at all- and only 19 that made any distinction- with either one being dashed or a different color, or one not being drawn at all. The designs that made a distinction tended to also have numbered or colored labels, and / or a key, seemingly to go with an instruction set. Designs without a distinction, however, were more likely to be the kind without any instructions as they are either intuitive enough to do without, or made for slightly more experienced papercrafters who already know the basics.

I think that the reason behind there being a distinction between fold and cut lines comes down to both the complexity of a design, and the assumed experience of the end user. Since most of our designs are composed of simple cuboids which are assembled essentially the same, people will know to make the same cuts and same folds where always are made, and therefore only places outside the regular cuboid would need to define what a cut and what a fold looks like.

However, if a distinction *must* be made, it should be in the lines' colors instead of dash pattern or lack thereof- this way it is both familiar and functional, and keeps the aforementioned benefits of a dashed line over a fold line, including that `drawFoldLineCuboid()` only needs one extra rectangle drawn over the center to make the change.

In short:

- Cut and fold lines being different tend to be present only when instructions are needed- for complex designs, or for less experienced papercrafters.
- Our designs tend to fall into the more intuitive category, and in line with sites with little to no instructions or labels, should be fine having no distinctions between cut and fold lines. 
- While most of the time lines without distinction are solid, it shouldn't hurt to keep using dashed ones to maintain how the site already uses them.

Sorry if some of these make no sense, but here's a few of my thoughts. Thanks!

**Comment by pixelpapercraft** [comment]
Hi @NinjolasNJM thanks for your detailed reply, really helped to clarify some of your thinking. 

> What I mean is that when a one in-game pixel wide edge

Ah, I see. TBH I am not concerned about those at all. On 1 pixel parts I would be surprised if anyone couldn't figure out to fold or not if the fold dash was not visible (excluding some edge cases). A fold line is about where to fold, but mostly it's a guide to help you get a straight fold on larger sections. Doing a clean fold without a fold line guide on a large section is hard to do by hand. So let's optimise our fold lines for that, don't be too worried about folds for 1 pixel size parts. Also, I'm happy to experiment with pattern sizes and see how they work.

> An edge with a solid border will look different from one with just a dashed one

I'm probably not following exactly the scenario you have in mind, but I am sure it can happen. Maybe that's why lots of designs just use straight lines everywhere? So let's say this point favours consistent line drawing everywhere, whether it be solid lines everywhere, or dashed lines everywhere - is that right? I assume not many people are actually concerned about this, however for those that might be it's the reason I originally implemented turning off fold lines so super clean edges everywhere without lines/dashes, so that's a practice worth continuing.

> other sites seem to be more disorganized both in their design conventions and with their users

100%, most of them are a mess, don't worry I have no intention of becoming like them. When I referred to the "greater papercraft community" that's not very helpful. What I had in mind was that we follow practices of the professional papercraft designers. Pick people, sites, books, etc. that you consider professional and let's use them as inspiration. Does not mean we literally copy them, but at least look at their practices and consider doing the same.

> From what I've researched,

Thanks for taking time to research that. Let's choose some references that we consider the best designers etc and see what they are all doing. I did a quick Google search now, but there is so much poor content out there. I'll look a bit more closely on the weekend.

> In short:

Great summary. I'm not really sold on the idea that we keep using dashed lines primarily because that's what we've done in the past. TBH I should have had this chat with you and others a long time ago. I have no idea where that dashed outline idea came from - it was probably something I did in a generator. It's not terrible, but it could have been better.

Let's say our goal is to come up with the best possible standard we can from now, don't worry about prior conventions. Also, it's just the edge lines that are affected, they actual textures remain unchanged so I suspect most people won't even notice.

Thanks for your amazing attention to detail and care for quality, I really appreciate the discussion.

In the meantime I'll try find some great reference designers. Appreciate if you have a chance to look into this as well. I might not have time during the week, but can take a look on the weekend.

**Comment by pixelpapercraft** [comment]
Hi @NinjolasNJM

> If updating standards is now the topic, then something big to consider is changing the standard DPI and/or resolution. 

It's a great idea, and something that I've thought about from time to time. There are some pros and cons to it, but that's something we can chat about. The way to do this properly is to move away from pixel based measurements to mm or cm based measurements. E.g. When drawing a texture we choose say an 8 x 8 pixel region and then draw as a 3cm x 3cm image. This idea works much better with print resolutions, and then it doesn't matter if the resolution is 75dpi or 300dpi or 600dpi, etc. Would have been better if I did this from the beginning, but that wasn't the trend back when I started the generators. Technically we could hack the current code to support a 300dpi resolution, and it's quite a lot of work, but the only noticeable difference would be sharper lines, so it would be better to do it properly. Also, we could consider a Generator V2 and we rethink the whole API for generator developers.

> I pushed a testing page with a couple of different patterns,

Great to see all of those variations for comparison.

A variation that would be great to see is solid lines for folds and cuts - that was a common pattern I noticed others using (I think this is the style that @PaperDoggy uses?)

> I asked the discord about this a few days ago

Great. Interesting to see such varied views. I'll look forward to hearing more about what you discover.

**Comment by NinjolasNJM** [comment]
>A variation that would be great to see is solid lines for folds and cuts - that was a common pattern I noticed others using (I think this is the style that @PaperDoggy uses?)

I pushed another testing page, please take a look.

> When drawing a texture we choose say an 8 x 8 pixel region and then draw as a 3cm x 3cm image.

The problem with doing this is that whatever is a generator standard also has to work as a manually made design standard as well. Using centimeters as the main measurement for the canvas would make design making on a number of design softwares much more difficult, as it's a lot easier to draw thinking about the pixels rather than how many centimeters the print will turn out to be. When printing, the same design will often not end up the same size anyways- measurements in metric when printers and their settings will confound them become misnomers. This also causes problems for generators- there would need to be a `rectangle2`, `position2` and `cuboid2` that all use floats to represent decimal cm values instead of whole pixel values.

I think that what would work best as far as introducing a new scale would be to have both the current 800% scale an a new, say 2400% scale have an a4 template page, and be supported in the generator builder. That way, not all generators break at once, and can be converted over if need be. Here, it would also be helpful to add a single, optional scalar parameter to drawImage() so that the same images can be used for converted generators. But this is all for a different PR.

At the current moment, I think we should:

- merge this PR if there isn't anything major that should be added or removed, keeping the current standards for now, and
- plan a larger potential design standards update for here and the site, maybe getting a few more voices involved, like M16 or perhaps a site veteran, like Maki if we can

Thanks for being open to all these ideas. These are really, really intriguing topics, but I feel like there are so many possibilities in them, and we might not be able to decide everything about them in this chat, so we should try and move it to somewhere where we can discuss it more in depth. What do you think?

**Comment by pixelpapercraft** [comment]
> The problem with doing this is that whatever is a generator standard also has to work 

Great feedback, definitely some more things to think about 👍 

> At the current moment, I think we should:

Yes, keen to get some of this merged. It's a bit too big to merge as is (and a few discussions incomplete). Would you mind opening new PRs with the following:

[1] fillRect() function.
[2] drawCuboid() function.

I have minor feedback that I'll provide on the PRs, but otherwise they look great. Once those are in then let's move onto some of the other functions.

I've pushed up a starter for the new `Minecraft` module which includes definitions for the cuboid types. Can you use these types in your code, and feel free to start adding new code there as well. Plus let me know if you have any feedback on that module.

Also, feel free to keep this PR open for now - you have done a lot of work here and we can take pieces of it for smaller PRs.

> so we should try and move it to somewhere where we can discuss it more in depth. What do you think?

Great idea. Were you thinking about a Discord channel, or something else?
